### ISSUE TITLE
Macro crashes with SIGSEGV when parameter is object

### ISSUE THREAD

**Issue by maxmilton** [oven-sh/bun#3832]
### What version of Bun is running?

0.7.0

### What platform is your computer?

Linux 6.4.6-arch1-1 x86_64 unknown

### What steps can reproduce the bug?

1. Create file `macro.ts`:
```ts
export function fn({ key }) {
  return key;
}
```
2. Create file `index.test.ts`;
```ts
import { expect, test } from 'bun:test';
import { fn } from './macro' assert { type: 'macro' };

test('segfaults', () => {
  expect(fn({ key: true })).toBe(true);
});
```
3. Run `bun test`.

### What is the expected behavior?

Test passes and bun process runs to completion without a segfault when object ~~destructuring~~ is used in macro arguments.

### What do you see instead?

```
~/P/_/bun-macro-sigsegv ❱ bun test
bun test v0.7.1 (f3200ac0)

index.test.ts:
fish: Job 1, 'bun test' terminated by signal SIGSEGV (Address boundary error)
```

### Additional information

- ~~Passing in an object _without_ destructuring in the macro function parameters works as expected.~~ See comment below.
- Passing an array works without error.
- Same result in both bun stable `v0.7.0 (aa1ad7f0)` and canary `v0.7.1 (f3200ac0)`.
- Both `*.ts` and `*.js` are also the same result.

**Comment by maxmilton** [comment]
Actually, it looks like I can also reproduce this without destructuring. E.g., changing `macro.ts` to:

```ts
export function fn(opts) {
  return opts.key;
}
```

Then running the above test will still segfault.

---

Also, interestingly, using an object (with or without destructuring) doesn't segfault if you don't set an object property that's accessed.

```ts
import { expect, test } from 'bun:test';
import { fn } from './macro' assert { type: 'macro' };

test('works', () => {
  expect(fn({ foo: true })).toBe(undefined);
});
```

Running this test will pass as expected. I assume it's due to some execution optimisation.

**Comment by maxmilton** [comment]
I also found doing object property access in the macro arg also fails, but only when using an object referenced by a variable:

```ts
import { expect, test } from 'bun:test';
import { fn } from './macro' assert { type: 'macro' };

test('works', () => {
  expect(fn({ text:'A' }.text)).toBe('A');
});

test('segfaults', () => {
  const obj = { text:'A' };
  expect(fn(obj.text)).toBe('A');
});
```

I think this is a regression in bun `v0.7.2`, as now my project tests are failing where env vars are passed into macros (e.g., `process.env.APP_VERSION`), but this previously worked.

**Comment by eye-wave** [comment]
i have another example of a macro causing SEGFAULT, but with a callback instead

### Bun version
1.0.7

### Operating System
Linux 6.5.6-76060506-generic x86_64 x86_64


```ts
// macro.ts
export const $macro =<T>(callback:() => T):T => callback()



// index.ts
import { $macro } from "./macro.ts" assert { type: "macro" }

const DATE = $macro(() => new Date().getTime())

console.log(DATE)
```

**Comment by eye-wave** [comment]
> ### What version of Bun is running?
> 
> 0.7.0
> 
> ### What platform is your computer?
> 
> Linux 6.4.6-arch1-1 x86_64 unknown
> 
> ### What steps can reproduce the bug?
> 
> 1. Create file `macro.ts`:
> ```ts
> export function fn({ key }) {
>   return key;
> }
> ```
> 2. Create file `index.test.ts`;
> ```ts
> import { expect, test } from 'bun:test';
> import { fn } from './macro' assert { type: 'macro' };
> 
> test('segfaults', () => {
>   expect(fn({ key: true })).toBe(true);
> });
> ```
> 3. Run `bun test`.
> 
> ### What is the expected behavior?
> 
> Test passes and bun process runs to completion without a segfault when object ~~destructuring~~ is used in macro arguments.
> 
> ### What do you see instead?
> 
> ```
> ~/P/_/bun-macro-sigsegv ❱ bun test
> bun test v0.7.1 (f3200ac0)
> 
> index.test.ts:
> fish: Job 1, 'bun test' terminated by signal SIGSEGV (Address boundary error)
> ```
> 
> ### Additional information
> 
> - ~~Passing in an object _without_ destructuring in the macro function parameters works as expected.~~ See comment below.
> - Passing an array works without error.
> - Same result in both bun stable `v0.7.0 (aa1ad7f0)` and canary `v0.7.1 (f3200ac0)`.
> - Both `*.ts` and `*.js` are also the same result.

it has been somewhat patched in 1.8. And instead of segfualt it just gives you a warning and doesn't run the code either way

**Comment by Jarred-Sumner** [comment]
Fixed as of Bun v1
